hibernate-core/src/main/java/org/hibernate/event/internal/DefaultFlushEntityEventListener.java

```diff
@@ -470,6 +473,21 @@ protected final boolean isUpdateNecessary(FlushEntityEvent event) throws Hiberna
 
 			int[] dirtyProperties = event.getDirtyProperties();
 			if ( dirtyProperties != null && dirtyProperties.length != 0 ) {
+				ExecutableList<AbstractEntityInsertAction> insertions = event.getSession().getActionQueue().getInsertions();
+				if ( insertions != null ) {
+					Iterator<AbstractEntityInsertAction> it = insertions.iterator();
+					while ( it.hasNext() ) {
+						AbstractEntityInsertAction insertAction = it.next();
+						if ( insertAction.getInstance() == event.getEntity() ) {
+							Object[] oldState = insertAction.getState();
+							Object[] newState = event.getPropertyValues();
+							for ( int i = 0; i < dirtyProperties.length; i++ ) {
+								oldState[dirtyProperties[i]] = newState[dirtyProperties[i]];
+							}
+							return false;
+						}
+					}
+				}
```

Comment:
would be great if `final` is used as aggressively as in other locations in Hibernate codebase.